### DISCUSSION TITLE
Can the Client Secret Safely Be Public?

### DISCUSSION THREAD

**Discussion by zicklag** [community/community#54568] [Apps, API and Webhooks] [answered]
### Select Topic Area

Question

### Body

Hey there! I'm working on an app that will deploy GitHub pages sites through the API.

To authenticate the user I'm wanting to use the OAuth web-app flow, but I need clarification about the security requirements of the `client_secret`. This app operates through a client-side app, that the user can have full access to the internals of, but in order to obtain an access token, I need the `client_secret` assocated to my OAuth app.

Is it safe to embed that `client_secret` with my ( open source ) client-side app where anybody could get to it?

**Comment by GeoffreyFClark** [comment]
Unfortunately it's not safe to embed the client_secret in a client-side app or open-source codebase where anybody can access it. 

A solution would be to use the Proof Key for Code Exchange (PKCE) extension.

**Reply by zicklag** [reply, marked as answer]
It appears that's not supported from the [GitHub docs](https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/authorizing-oauth-apps#1-request-a-users-github-identity):

> The PKCE (Proof Key for Code Exchange) parameters code_challenge and code_challenge_method are not supported at this time.

---

It looks like the [Device Flow](https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/authorizing-oauth-apps#device-flow) would be a good substitute, though.

Slightly less convenient because the user has to input the code, but at least it can be done fully client-side without including the client secret.

**Comment by zicklag** [comment]
Does anybody know if there is a way to auto-fill the [device activation page](https://github.com/login/device) with a code through a query string parameter? That would make the workflow just about as friendly as the web app flow.
